stream edit: Add support for adding all members of another stream.

ryanreh99 opened the pull request:
#15186
WIP PR because the non-typeahead  case (user presses the `Esc` key)
has not been implemented yet, will add it in a new commit.
And the logic for user-groups is remaining too, will create a new PR for this.

@shubhamdhama @pragatiagrawal31 could you have a look.

**Testing Plan:**
Added the node tests for all changes made here.

shubhamdhama commented on static/js/user_pill.js:
```diff
@@ -146,7 +148,9 @@ exports.set_up_typeahead_on_pills = function (input, pills, update_func, source)
             exports.append_user(user, pills);
             input.trigger("focus");
-            update_func();
+            if (opts.update_func && typeof opts.update_func === "function") {
```
Instead of doing this function-check in `if` statement I'd have done something like this,
```js
            if (opts.update_func) {
                if (opts.update_func !== "function") {
                    blueslip.error(<error message here>);
                }
                opts.update_func();
            }
```
But I'm not sure we are doing these sort of checks at other places. But it's definitely a good thing to do. Maybe, typescript will take care of this in future.

shubhamdhama replied:
A small commit message for this commit 9cec9ecfd41dc1d2204b36406c4eb5190a9be66f is appreciated. Can be 1-3 lines explaining that it's a preliminary commit.

ryanreh99 replied:
> But I'm not sure we are doing these sort of checks at other places

I see, I checked at other typeahead places and this check wasn't necessary as javascript already raises a `TypeError`.

shubhamdhama commented:
We should show the full names instead? 
![image](https://user-images.githubusercontent.com/22238472/88174682-4eca2580-cc42-11ea-9103-ac5e0317d627.png)

ryanreh99 commented:
Thanks for the reviews.

> Instead of extending `set_up_typeahead_on_pills`, try sending those special cases on the calling side to keep this function clean.

Oh I implemented this mostly according to the composebox typeahead (or rather how it was before commit [b65da7c](https://github.com/zulip/zulip/commit/b65da7cbe9536b8a7dc8f6dab3c61c67b8da3f3b)).
And was planning on adding another condition for the user-groups case too.

This is currently used in only one place, but I was thinking this could be useful if we decide to use this logic elsewhere.
Such as, the `#organization/user-groups-admin` page.


> We should show the full names instead?

I agree, I'll add this as a followup.

timabbott commented:
I merged the first commit.  I feel a bit uncomfortable adding this feature via changes to `user_pill.js`; that library is pretty clearly intended to be user pills.  Is there a better way we can add the feature while preserving the logical integrity of "user_pill" module?  Might require some nontrivial refactoring; conceptually we probably want a `user_pill.js` and a `stream_pill.js` and perhaps a new `pill_typeahead.js` that supports user and stream pills?  Not sure.

timabbott commented:
> We should show the full names instead?

Yeah and we should do the same thing we do in "User settings" to show the user's details.

ryanreh99 commented:
@timabbott @shubhamdhama .
This includes the remaining work for this feature (the non-typeahead use case).

Now the only user_pill changes are extracting out the typeahead setup.
Moved the stream and user pills combining logic to the calling side in `submit_add_subscriber_form`.

Added the placeholder text in the last commit as I think this text could be improved.
Or we could add a `(?)` button beside `Stream membership`, which links to a User Documentation page.

timabbott commented:
Can you reorder the earlier commits to do the extractions to a new file before adding the new feature?  It'll be a lot easier for me to follow that way.  More generally I don't understand why the "Add test file" commit has so much other refactoring in it; can you do a bit of work on making this branch readable?

ryanreh99 commented:
@timabbott, I Updated this.
Set the pill typeahead extraction as the first commit.

And removed the unnecessary refactoring from the `stream_edit` test file commit.
The only 2 changes made for this are necessary:
• the `get_stream_id ` function is exported so that we can override it.
• We calculate stream id in `show_subscription_settings` from the sub object instead of the DOM element.

timabbott commented:
I merged the first commit as 60eef68f5974664af2f6a0dc5bbbe42aea607393.  You'll want to rebase and run `lint`; new modules need `use strict` added at the top.

timabbott commented on static/js/stream_pill.js:
```diff
@@ -0,0 +1,51 @@
+            continue;
+        }
+        const subscribers = stream_edit.get_users_from_subscribers(sub.subscribers);
+        user_ids = user_ids.concat(subscribers.map((user) => user.user_id));
```
These two lines are weird -- `sub.subscribers` is already a list of user ids; so you want to just `user_ids = user_ids.concat(sub.subscribers)`.

Also should this be called `get_user_ids_from_sub`, since it's potentially multiple streams?

timabbott commented on static/js/stream_pill.js:
```diff
@@ -0,0 +1,51 @@
+    pill_widget.appendValidatedData({
+        display_value: "#" + stream.name + ": " + stream.subscriber_count + " users",
+        stream_id: stream.stream_id,
+    });
```
When displaying these pills, do we want to show the number of users being added?  I think that'd be a very helpful visual hint.

timabbott replied:
Oh, I see, this is in a later commit.  OK.

timabbott commented on static/js/stream_edit.js:
```diff
@@ -211,9 +211,18 @@ function submit_add_subscriber_form(e) {
+        // case occurs when creating user pills from a stream.
+        const index = user_ids.indexOf(page_params.user_id);
+        user_ids.splice(index, 1);
+    }
```
Couldn't this be done much more simply while we still have a `Set` by just removing the current user ID, rather than doing this awkward array splicing?

timabbott commented on frontend_tests/node_tests/stream_edit.js:
```diff
@@ -0,0 +1,271 @@
+    user_pill.get_user_ids = () => [mark.user_id, me.user_id];
+    add_subscribers_handler(event);
+    assert(called);
+});
```
I feel like this test is too mock-heavy.  Ideally, we should be doing end-to-end tests of the main logic, e.g. "Type these 5 characters, verify the typeahead options are X, Y, Z in that order".  Then "Select Y, verify it creates a pill with Z data".  Reading this test it seems like all we're doing is getting test "coverage" -- i.e. checking all the lines execute -- but not verifying much about the actual output.

(I could be missing something as I'm mostly skimming these tests)

https://zulip.readthedocs.io/en/latest/testing/philosophy.html

@showell or @aero31aero may be able to provide more advice on setting this test up properly; our main typeahead tests balance this stuff well.

timabbott commented on static/js/stream_edit.js:
```diff
@@ -120,16 +120,16 @@ function clear_edit_panel() {
     const row = target.closest(".stream-row, .subscription_settings");
     return parseInt(row.attr("data-stream-id"), 10);
-}
+};
```
If we're exporting this, we should add a comment defining its interface -- what does it do when called if one doesn't have the stream-edit UI open?

timabbott commented on static/js/stream_edit.js:
```diff
@@ -292,9 +292,8 @@ exports.sort_but_pin_current_user_on_top = function (users) {
-    const stream_id = sub_row.data("stream-id");
-    const sub = stream_data.get_sub_by_id(stream_id);
+function show_subscription_settings(sub) {
+    const stream_id = sub.stream_id;
```
The refactor to pass this a `sub` might be worth doing as a prep commit to simplify things here.

timabbott commented on static/js/stream_edit.js:
```diff
@@ -306,10 +306,26 @@ function show_subscription_settings(sub) {
+            return text;
+        }
+        return user_pill.get_email_from_item(item);
+    }
```
I don't really understand what this commit does; is this interface for copy-pasting content into the box?

ryanreh99 replied:
Yes besides copy pasting, this last commit is added so that stream pills can be created when the user does not hides the typeahead (presses `Esc` key before enter). The user pill also has this feature.

( I updated the commit )

timabbott commented on frontend_tests/node_tests/stream_pill.js:
```diff
@@ -0,0 +1,45 @@
+    test_create_item("sweden", [], undefined);
+    test_create_item("#sweden", [sweden_pill], undefined);
+    test_create_item("  #sweden", [], sweden_pill);
+    test_create_item("#test", [], undefined);
```
This sort of test is great.

timabbott commented on static/templates/subscription_members.hbs:
```diff
@@ -12,7 +12,7 @@
                     <div class="pill-container person_picker">
                         <div class="input" contenteditable="true"
-                          data-placeholder="{{t 'Add subscribers' }}"></div>
+                          data-placeholder="{{t 'Enter users or prepend # to a stream to add subscribers' }}"></div>
```
Maybe "Add subscribers.  Use #streamname to subscribe a whole stream".

timabbott commented:
I posted a long batch of comments.  Overall this is quite promising; just have a few code details to polish.

ryanreh99 commented:
@timabbott I resolved the review comments and this is ready for review.

Now, in the `stream_edit.js` test file, mainly the DOM elements are stubbed.
I also mocked `stream_pill.get_user_ids` and `user_pill.get_user_ids` after the first event handler verifies these functions.
Then the rest of the event handlers uses these mocked functions to verify the user ids that are going to be subscribed based on the current pills created.
The typeahead functions tests are done similar to that in `settings_user_groups.js`.

timabbott commented:
This is great, merged, thanks @ryanreh99!  I guess for #15186 we still need to do something similar for user groups?

One thing we should fix as a quick follow-up is the success output after adding a stream:

![image](https://user-images.githubusercontent.com/2746074/91779664-14419a00-ebab-11ea-8260-e322ff400efb.png)

Ideally, we should provide names (with links to profiles) rather than emails, and also we should probably list the users who were successfully added in this "we added about half the users from your request" case, not just those who weren't.

timabbott commented:
@ryanreh99 can you open an issue or PR for that follow-up?
